File: The K-means algorithm/kmeans.py
import sys

def kmeans(k ,num_of_iter ,file):
    try:
        num_from_file=[]
        clusters = []
        centroides = []
        with open(file) as file:
            for line in file:
                arr = line.rstrip().split(",")
                for i in range(0, len(arr)):
                    arr[i] = float(arr[i])
                num_from_file.append(arr)
        N = len(num_from_file)
        error = False
        if k<=1 or k>=N or k!=int(k):
            print("Invalid number of clusters!")
            error = True
        if num_of_iter<=1 or num_of_iter>=1000 or num_of_iter!=int(num_of_iter) :
            print("Invalid maximum iteration!")
            error = True
        if (error):
                return

        for i in range(0,k):
            centroides.append(num_from_file[i])
            
        initialize_clusters(clusters,k)
        curr_num_of_iter = 0
        while(curr_num_of_iter < num_of_iter):
            curr_num_of_iter += 1
            initialize_clusters(clusters,k)
            refill_cluster(num_from_file,centroides,clusters)
            stop = update_centroides(centroides,clusters)
            #print(" cluster = ",clusters ," centroides = ", centroides)
            if (stop):
                break
            
        #print centroides
        for i in range (0,len(centroides)):
            for j in range (0,len(centroides[i])):
                if (j < len(centroides[i])-1):
                    print("%.4f" %centroides[i][j],end=",")
                else:
                     print("%.4f" %centroides[i][j])
                     
        
    except ValueError :
        print("An Error Has Occured")

        
                
    
def initialize_clusters(clusters,k):
    clusters.clear()
    for i in range (0,k):
        clusters.append([])


def refill_cluster(num_from_file,centroides,clusters):
    positive_infinity = float('inf')
    for i in range (0 , len(num_from_file)):#iterate through the elemnts in num from file
        min_distance = positive_infinity
        index_of_min_distance = -1
        for j in range (0 , len(centroides)):#iterate through centoides
            curr_min = compute_distance(num_from_file[i],centroides[j])
            if (curr_min < min_distance):
                min_distance = curr_min
                index_of_min_distance = j
        clusters[index_of_min_distance].append(num_from_file[i])
    
    
def compute_distance(vector1 ,vector2):
    tmp = 0
    for i in range (0 , len(vector1)):
        tmp += pow((vector1[i] - vector2[i]),2)
    tmp = pow(tmp,0.5)
    return tmp

def update_centroides(centroides,clusters):
    stop = True
    for j in range (0 , len(centroides)):#iterate through centoides
        new_centroid =[]
        for t in range (0,len(centroides[j])): #iterate through coordinates
            tmp = 0
            for i in range (0,len(clusters[j])):#iterate through elemnts in cluster
                tmp += clusters[j][i][t]
            tmp = tmp/(len(clusters[j]))
            new_centroid.append(tmp)
        if ( compute_distance(new_centroid,centroides[j]) >= 0.001 ):
            stop = False
        centroides[j] = new_centroid
    return stop

input_argv=sys.argv
input_argc=len(sys.argv)
try:
    k=int(input_argv[1])
    
except ValueError :
        print("Invalid number of clusters!")
        
try:
    if (input_argc==4):
        iteration = int(input_argv[2])
        kmeans(k,iteration,(input_argv[3]))

    elif (input_argc==3):
        kmeans(int(input_argv[1]),200,(input_argv[2]))
        
    else:
        print("An Error Has Occured")
        
except ValueError :
        print("Invalid maximum iteration!")

File: The K-means algorithm/kmeans.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

double compute_distance(double* vector1, double* vector2, int d);
void refill_cluster(double** centroids, int** clusters, double** num_from_file, int K, int N, int d);
int update_centroides(int** clusters, double** centroids, double** num_from_file, int K, int d);
void append_to_cluster_j(int** clusters, int j, int index);
int** initialize_clusters(int K);
void kmeans(int K, int iter, char* input);

int main(int argc, char** argv)
{
    int k = strtol(argv[1], NULL, 10);
    if (argc >= 4)
    {
        int iter = strtol(argv[2], NULL, 10);
        kmeans(k, iter, argv[3]);
    }
    else
    {
        kmeans(k, 200, argv[2]);
    }
    return 0;

}

void kmeans(int K, int iter, char* input)
{
    double** centroids =NULL;
    double** num_from_file = NULL;
    int i=0;
    int j=0;
    int d = 0;
    int s = 0;
    int e = 0;
    int num_of_curr_iter = 0;
    FILE* file_p;
    char* line = NULL;
    ssize_t current;
    int stop=0;
    size_t length = 0;
    int num_lines = 0;
    int dimensions = 1;
    int N = 0;
    int** clusters=NULL;
    int invalid_k = 0, invalid_iter = 0;

    file_p = fopen(input, "r");
    if (file_p == NULL)
    {
        perror("An Error Has Occurred");
        exit(1);
    }

    while ((current = getline(&line, &length, file_p)) != -1)
    {
        num_lines++;
        if (num_lines == 1)
        {
            int j = 0;
            while (line[j] != '\0')
            {
                if (line[j] == ',')
                {
                    dimensions++;
                }
                j++;
            }
        }
    }

    
    N = num_lines;
    fseek(file_p, 0, SEEK_SET);

    if (K >= N || floor(K) != K || K < 1)
    {
        fprintf(stderr, "Invalid number of clusters!\n");
        invalid_k = 1;
    }
    if (iter >= 1000 || iter < 1 || floor(iter) != iter)
    {
        fprintf(stderr, "Invalid maximum iteration!\n");
        invalid_iter = 1;
    }
    if (invalid_k || invalid_iter)
    {
        exit(1);
    }
    fseek(file_p, 0, SEEK_SET);
    num_from_file= (double**)malloc(num_lines * sizeof(double*));
    if(num_from_file == NULL){
        perror("An Error Has Occurred");
        exit(1);
    }
    for ( i = 0; i < num_lines; i++)
    {
        double* arr = (double*)malloc(dimensions * sizeof(double));
        if(arr == NULL){
            perror("An Error Has Occurred");
            exit(1);
        }
        current = getline(&line, &length, file_p);
        s=0;
        e=0;
        j=0;
        while (e <= current)
        {
            if (line[e] == ',' || line[e] == '\0')
            {
                line[e] = '\0';
                arr[j++] = atof(&line[s]);
                s = e + 1;
            }
            
            e++;
        }
        num_from_file[i] = arr;
    }


    centroids= (double**)malloc(K * sizeof(double*));
    if(centroids == NULL){
        perror("An Error Has Occurred");
        exit(1);
    }
    for ( i = 0; i < K; i++)
    {
        centroids[i] = num_from_file[i];
    }

    d = dimensions;
    num_of_curr_iter = 0;
    
    clusters= initialize_clusters(K);
    while (num_of_curr_iter < iter)
    {  
        stop=0; 
        refill_cluster(centroids, clusters, num_from_file, K, N, dimensions);
        stop = update_centroides(clusters, centroids, num_from_file, K, dimensions);
        if (stop)
        {
            break;
        }
        num_of_curr_iter++;

        for ( i = 0; i < K; i++)
        {
            free(clusters[i]);
        }
        clusters = initialize_clusters(K);
        if(clusters == NULL){
            perror("An Error Has Occurred");
            exit(1);
    }
    }

    fclose(file_p);
    if (line)
    {
        free(line);
    }

    for ( i = 0; i < K; i++)
    {
        for ( j = 0; j < d; j++)
        {
            if(j<d-1)
                printf("%0.4f,", centroids[i][j]);
            else
                printf("%0.4f", centroids[i][j]);
        }
        printf("\n");
    }

    for ( i = 0; i < num_lines; i++)
    {
        free(num_from_file[i]);
    }
    free(num_from_file);
    free(centroids);
}

void append_to_cluster_j(int** clusters, int j, int index)
{
    int* tmp=NULL;
    clusters[j][0]++;
    tmp = (int*)realloc(clusters[j], (clusters[j][0] + 1) * sizeof(int));
    if (tmp != NULL)
    {
        clusters[j] = tmp;
        clusters[j][clusters[j][0]] = index;
    }
    else{
        perror("An Error Has Occurred");
        exit(1);
    }


}

int** initialize_clusters(int K)
{
    int** clusters;
    int i=0;
    int** tmp = (int**)calloc(K, sizeof(int*));
    int* tmp_1=NULL;
    if (tmp != NULL)
    {
        clusters = tmp;
        for ( i = 0; i < K; i++)
        {
             tmp_1 = (int*)calloc(1, sizeof(int));
            if (tmp_1 != NULL)
            {
                clusters[i] = tmp_1;
                clusters[i][0] = 0;
            }
            else{
                perror("An Error Has Occurred");
                exit(1);
    }
        }
    }
    else{
        perror("An Error Has Occurred");
        exit(1);
    }
    tmp_1[0]=0;
    tmp[0][0]=0;
    return clusters;
}

int update_centroides(int** clusters, double** centroids, double** num_from_file, int K, int d)
{
    int stop = 1;
    int m=0;
    int j=0,i=0;
    double* tmp=NULL;
    double* old_centroid = NULL;
    old_centroid=centroids[0];
    for ( i = 0; i < K; i++)
    {
        old_centroid = centroids[i];
        tmp = (double*)calloc(d, sizeof(double));
        if (tmp != NULL)
        {
            int num_of_elemnt_in_cluster = clusters[i][0];
            for ( j = 1; j <= num_of_elemnt_in_cluster; j++)
            {
                double* vector = num_from_file[clusters[i][j]];
                for ( m = 0; m < d; m++)
                {
                    tmp[m] += vector[m];
                }
            }
            for ( j = 0; j < d; j++)
            {
                tmp[j] = tmp[j] * (1.0 / num_of_elemnt_in_cluster);
            }
            if (compute_distance(tmp, old_centroid, d) >= 0.001)
            {
                stop = 0;
            }
            centroids[i] = tmp;
        }
        else{
            perror("An Error Has Occurred");
            exit(1);
         }
    }
    return stop;
}

void refill_cluster(double** centroids, int** clusters, double** num_from_file, int K, int N, int d)
{
    int i=0,j=0;
    for ( i = 0; i < N; i++)
    {
        double* vector = num_from_file[i];
        int cluster_index = 0;
        double min_distance = INFINITY;
        for ( j = 0; j < K; j++)
        {
            double distance = compute_distance(vector, centroids[j], d);
            if (distance < min_distance)
            {
                min_distance = distance;
                cluster_index = j;
            }
        }
        append_to_cluster_j(clusters, cluster_index, i);
    }
}

double compute_distance(double* vector1, double* vector2, int d)
{
    int i=0;
    double distance = 0.0;
    for ( i = 0; i < d; i++)
    {
        distance += pow(vector1[i] - vector2[i], 2);
    }
    return sqrt(distance);
}
